src/lib.rs

```diff
@@ -0,0 +1,1058 @@
+///     The degree of the right polynomial.
+/// vars : sequence of :class:`MulVar`
+///     List of variables. For each output variable ``vars`` specifies if the
+///     variable exists in the left polynomial, the right polynomial or both.
+///
+/// Example
+/// -------
+///
+/// Consider the following polynomials:
+///
+/// .. math:: f(x) = x^2 + 2 x + 3
+/// .. math:: g(x) = 3 x + 1
+///
+/// The coefficients for the product :math:`p(x) = f(x) g(x)`:
+///
+/// >>> import numpy
+/// >>> plan = MulPlan(2, 1, [MulVar.Both])
+/// >>> numpy.testing.assert_allclose(
+/// ...     plan(numpy.array([1, 2, 3], dtype=float), numpy.array([3, 1], dtype=float)),
+/// ...     [3, 7, 11, 3],
+/// ... )
+///
+/// The coefficients for the product :math:`q(x, y) = f(x) g(y)`:
+///
+/// >>> import numpy
+/// >>> plan = MulPlan(2, 1, [MulVar.Left, MulVar.Right])
+/// >>> numpy.testing.assert_allclose(
+/// ...     plan(numpy.array([1, 2, 3], dtype=float), numpy.array([3, 1], dtype=float)),
+/// ...     [0, 0, 0, 3, 6, 9, 0, 1, 2, 3],
+/// ... )
+///
+/// See also
+/// --------
+/// :func:`mul` : Compute the multiplication without a plan.
+#[pyclass]
+#[pyo3(name = "MulPlan")]
+#[pyo3(
+    text_signature = "(degree_left, degree_right, vars)",
+    module = "nutils_poly"
+)]
+#[derive(Debug, Clone)]
+struct PyMulPlan(MulPlan);
+
+#[pymethods]
+impl PyMulPlan {
+    #[new]
+    fn new(degree_left: Power, degree_right: Power, vars: Vec<PyMulVar>) -> PyResult<PyMulPlan> {
+        Ok(Self(MulPlan::new(
+            degree_left,
+            degree_right,
+            &vars.as_sqnc().map(|var| var.0),
+        )))
+    }
+    fn __call__<'py>(
+        &self,
+        py: Python<'py>,
+        coeffs_left: PyReadonlyArrayDyn<f64>,
+        coeffs_right: PyReadonlyArrayDyn<f64>,
+    ) -> PyResult<&'py PyArrayDyn<f64>> {
+        let (coeffs_left, leading_shape_left) =
+            as_coeffs_dyn_with_ncoeffs(&coeffs_left, self.0.ncoeffs_left(), "coeffs_left")?;
+        let (coeffs_right, leading_shape_right) =
+            as_coeffs_dyn_with_ncoeffs(&coeffs_right, self.0.ncoeffs_right(), "coeffs_right")?;
+        let leading_shape = broadcast_shapes(&[leading_shape_left, leading_shape_right])?;
+        let coeffs_left = coeffs_left
+            .broadcast(shape![&leading_shape, [self.0.ncoeffs_left()]])
+            .unwrap();
+        let coeffs_right = coeffs_right
+            .broadcast(shape![&leading_shape, [self.0.ncoeffs_right()]])
+            .unwrap();
+        let result_shape = shape![&leading_shape, [self.0.ncoeffs_output()]];
+        let mut result = Vec::with_capacity(result_shape.iter().copied().product());
+        if has_contiguous_rows(&coeffs_left) && has_contiguous_rows(&coeffs_right) {
+            for (coeffs_left, coeffs_right) in iter::zip(coeffs_left.rows(), coeffs_right.rows()) {
+                let coeffs_left = coeffs_left.as_slice().unwrap().as_sqnc();
+                let coeffs_right = coeffs_right.as_slice().unwrap().as_sqnc();
```

Comment:
Not sure that I understand why `as_sqnc` is required here as we know that the result of `as_slice` implements `Sequence` (doesn't it?)

Reply:
The problem is with the reference. `MulPlan.apply` takes ownership over the arguments. `Sequence` is implemented for `[T]` (`T` generic), but not for `&[T]`.

Reply:
So, for my understanding, could we use `*coeffs_left` instead? (not suggesting we should)

Reply:
No, because `[T]` is unsized.

Reply:
I thought you were gonna say that :)